// @sanity/runtime-cli/dist/actions/blueprints/operations.js
import config from '../../config.js';
import getHeaders from '../../utils/get-headers.js';
const { apiUrl } = config;
export const stacksUrl = `${apiUrl}vX/blueprints/stacks`;
export async function getOperation({ stackId, operationId, auth, }) {
    const path = `${stacksUrl}/${stackId}/operations/${operationId}`;
    const response = await fetch(path, {
        method: 'GET',
        headers: getHeaders(auth),
    });
    if (!response.ok) {
        const errorText = await response.text();
        return {
            ok: false,
            error: errorText || 'Failed to fetch operation details',
            operation: null,
        };
    }
    const operation = await response.json();
    return {
        ok: true,
        error: null,
        operation,
    };
}
export async function listOperations({ stackId, auth, }) {
    const path = `${stacksUrl}/${stackId}/operations`;
    const response = await fetch(path, {
        method: 'GET',
        headers: getHeaders(auth),
    });
    if (!response.ok) {
        const errorText = await response.text();
        return {
            ok: false,
            error: errorText || 'Failed to fetch operations',
            operations: [],
        };
    }
    const operations = await response.json();
    return {
        ok: true,
        error: null,
        operations,
    };
}
